dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -0,0 +1,816 @@
+            throws InvalidConfigurationException {
+        final var analysis = new InclusionBasedPointerAnalysis(analysisContext.requires(Dependency.class));
+        analysis.run(program);
+        logger.debug("variable count: {}", analysis.totalVariables);
+        logger.debug("replacement count: {}", analysis.totalReplacements);
+        logger.debug("alignment sizes: {}", analysis.totalAlignmentSizes);
+        logger.debug("addInto: {} successes vs {} fails", analysis.succeededAddInto, analysis.failedAddInto);
+        return analysis;
+    }
+
+    private InclusionBasedPointerAnalysis(Dependency d) { dependency = d; }
+
+    // ================================ API ================================
+
+    @Override
+    public boolean mayAlias(MemoryCoreEvent x, MemoryCoreEvent y) {
+        final Offset<Variable> vx = eventAddressSpaceMap.get(x);
+        final Offset<Variable> vy = eventAddressSpaceMap.get(y);
+        if (vx == null || vy == null) {
+            return true;
+        }
+        if (vx.base == vy.base && vx.alignment.isEmpty() && vy.alignment.isEmpty()) {
+            return vx.offset == vy.offset;
+        }
+        final List<Offset<Variable>> ox = new ArrayList<>(vx.base.includes);
+        ox.add(new Offset<>(vx.base, 0, List.of()));
+        final List<Offset<Variable>> oy = new ArrayList<>(vy.base.includes);
+        oy.add(new Offset<>(vy.base, 0, List.of()));
+        for (final Offset<Variable> ax : ox) {
+            for (final Offset<Variable> ay : oy) {
+                if (ax.base == ay.base && overlaps(ay.offset + vy.offset - ax.offset - vx.offset,
+                        join(ax.alignment, vx.alignment), join(ay.alignment, vy.alignment))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean mustAlias(MemoryCoreEvent x, MemoryCoreEvent y) {
+        final Offset<Variable> vx = eventAddressSpaceMap.get(x);
+        final Offset<Variable> vy = eventAddressSpaceMap.get(y);
+        return vx != null && vy != null && vx.base == vy.base && vx.offset == vy.offset &&
+                vx.alignment.isEmpty() && vy.alignment.isEmpty();
+    }
+
+    // ================================ Processing ================================
+
+    private void run(Program program) {
+        checkArgument(program.isCompiled(), "The program must be compiled first.");
+        // Pre-processing:
+        for (final MemoryObject object : program.getMemory().getObjects()) {
+            totalVariables++;
+            constantMap.put(object, new Variable(object, object.toString()));
+        }
+        for (final RegWriter writer : program.getThreadEvents(RegWriter.class)) {
+            processWriter(writer);
+        }
+        for (final MemoryCoreEvent store : program.getThreadEvents(MemoryCoreEvent.class)) {
+            processMemoryEvent(store);
+        }
```

Comment:
The naming here is definitely odd. If this processes only stores, it should look for stores only. If not, the variable should have a different name.

Reply:
The method `processMemoryEvent` also processes `GenericMemoryEvent`, i.e. `synchronize_srcu`.  Currently, `Load` is the only special case, since it is both a writer and a memory event.  Hence, there is a comment in `processMemoryEvent`.
To have `Load`'s be processed before `Store`'s allowed me to omit the communication checks for `Load`.  Also, to have `Load`'s be in the same processing order as other writers avoids `replace(Variable,Offset<Variable>)`.
I could name it `processWritersExceptLoads`, if that helps reading the code.

Reply:
You can name it `processNonLoads/processStores` or similar but the name of the loop variable should match.
Currently, there is this odd discrepancy between the loop variable referring to stores specifically but the method seemingly accepting arbitrary memory events.
You could also make this loop filter out the loads rather than hiding it in the called method (the method could then assert that it excepts non-load events).